appcheck/firebase-appcheck/src/main/java/com/google/firebase/appcheck/internal/DefaultFirebaseAppCheck.java

```diff
@@ -137,33 +161,63 @@ public void removeAppCheckTokenListener(@NonNull AppCheckTokenListener listener)
               new FirebaseException("No AppCheckProvider installed.")));
     }
     // TODO: Cache the in-flight task.
+    return fetchTokenResultFromProvider();
+  }
+
+  @NonNull
+  @Override
+  public Task<AppCheckToken> getAppCheckToken(boolean forceRefresh) {
+    if (!forceRefresh && hasValidToken()) {
+      return Tasks.forResult(cachedToken);
+    }
+    if (appCheckProvider == null) {
+      return Tasks.forException(new FirebaseException("No AppCheckProvider installed."));
+    }
     return fetchTokenFromProvider();
   }
 
   /** Fetches an {@link AppCheckTokenResult} via the installed {@link AppCheckProvider}. */
-  Task<AppCheckTokenResult> fetchTokenFromProvider() {
-    return appCheckProvider
-        .getToken()
+  Task<AppCheckTokenResult> fetchTokenResultFromProvider() {
```

Comment:
Splitting this method in two seems like a code smell - the inner one is for Tokens, and the outer one is for TokenResults, but then you're still constructing TokenResults in the inner one in order to trigger the listener. Is there a reason (testing maybe?) that we've done it this way? I agree that splitting the listeners into the places where their insides are constructed would also be error-prone, which is why I wonder why we've done the split at all.

Reply:
The split is because the `getAppCheckToken` method needs to return an `AppCheckToken` while the `getToken` method needs to return an `AppCheckTokenResult`. The `fetchTokenFromProvider` method should do all the heavy-lifting for getting the token, updating the stored state, and notifying listeners. All `fetchTokenResultFromProvider` does is wrap the result into an `AppCheckTokenResult` for Firebase SDKs to consume.

Would it be more clear for me to get rid of the `fetchTokenResultFromProvider` method, and handle wrapping the token into an `AppCheckTokenResult` directly within `getToken`?

Reply:
It's just weird that you have:
- a place for marshalling things into TokenResults
- a place for internal heavy lifting, which also marshalls things into TokenResults
This would be resolved if you had a helper method that takes in a list of listeners of both types and a token and then marshals when necessary. Then you just have:
- a place that takes a token and marshalls it into a TokenResult
- a place that does everything internally, but only manipulates the Token itself. 
That being said, all we'd be doing is hiding the fact that we're duplicating work, so maybe that's not better either. I just know that the whole "have two Listener classes that are almost, but not quite, the same everywhere" is (IMO) just asking for a bug at some point where someone includes one but not both because we haven't isolated the pattern of dealing with both kinds of listeners to a single place. But that might be more trouble than it's worth, I leave it to you. I'm just trying to flag the concern - if I thought I had a clearly better way to do it, I'd say so, but I don't.

Reply:
Ack, for this PR I'll just inline the `AppCheckTokenResult` wrapping into `getToken` so we don't have two separate methods that do nearly the same thing.

In a follow-up PR, I'll do some refactoring to wrap the two lists of listeners into an internal class, so the logic to deal with the listeners can be encapsulated in one place.